app/controllers/api/api_controller.rb

```diff
@@ -2,68 +2,165 @@ class Api::ApiController < ApplicationController
   respond_to :json
 
   attr_accessor :current_user
+  attr_accessor :current_session
   attr_accessor :user_manager
 
   before_action :authenticate_user
   before_action :set_raven_context
 
   before_action do
     request.env['HTTP_ACCEPT_ENCODING'] = 'gzip'
-    self.user_manager = SyncEngine::V20190520::UserManager.new(User)
+    self.user_manager = SyncEngine::V20200115::UserManager.new(User)
   end
 
-  protected
+  private
 
   def authenticate_user
-    unless request.headers['Authorization'].present?
-      render_invalid_auth
+    token = token_from_request_header
+
+    if token.nil?
+      render_invalid_auth_error
       return
     end
 
-    strategy, token = request.headers['Authorization'].split(' ')
-    if (strategy || '').downcase != 'bearer'
-      render_invalid_auth
+    authentication = validate_token token
+
+    if authentication.nil?
+      render_invalid_auth_error
       return
     end
 
-    claims = begin
-               SyncEngine::JwtHelper.decode(token)
-             rescue
-               nil
-             end
-    user = User.find_by_uuid claims['user_uuid'] unless claims.nil?
+    user = user_from_authentication authentication
 
     if user.nil?
-      render_invalid_auth
+      render_invalid_auth_error
+      return
+    end
+
+    if authentication[:type] == 'jwt' && params[:api].to_i >= 20190520 && user.supports_sessions?
+      render_invalid_auth_error
+      return
+    end
+
+    if authentication[:type] == 'session_token' && authentication[:session].expired?
+      render_invalid_auth_error
+      return
+    end
+
+    if authentication[:type] == 'session_token' && authentication[:session].expired_access_token?
+      render_expired_token_error
       return
     end
 
     # If a user signed in before the JWT change was made below, they won't have a pw_hash.
-    if claims['pw_hash']
+    if authentication[:type] == 'jwt' && authentication[:claims]['pw_hash']
+      pw_hash = authentication[:claims]['pw_hash']
       encrypted_password_digest = Digest::SHA256.hexdigest(user.encrypted_password)
       # newer versions of our jwt include the user's hashed encrypted pw,
       # to check if the user has changed their pw and thus forbid them from access if they have an old jwt
-      if ActiveSupport::SecurityUtils.secure_compare(claims['pw_hash'], encrypted_password_digest) == false
-        render_invalid_auth
+      unless ActiveSupport::SecurityUtils.secure_compare(pw_hash, encrypted_password_digest)
+        render_invalid_auth_error
         return
       end
     end
 
     self.current_user = user
+    self.current_session = authentication[:session] if authentication[:type] == 'session_token'
+  end
+
+  def authenticate_user_for_sign_out
```

Comment:
This method is used to authenticate users when they try to `sign_out` only: https://github.com/standardnotes/syncing-server/blob/27308613e8acf531d4a26e3ae908fd7c82bc7eb8/app/controllers/api/auth_controller.rb#L3

Used for this case: https://github.com/standardnotes/snjs/pull/11#discussion_r423243854

Basically, it authenticates the user in order to terminate their session (even if the session is expired). 

We can not use: https://github.com/standardnotes/syncing-server/blob/27308613e8acf531d4a26e3ae908fd7c82bc7eb8/app/controllers/api/api_controller.rb#L18

Because the method has the check to see if a session is expired.

Reply:
Sounds good, is it worth commenting in the code?